### ISSUE TITLE
Missing Skip Link

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#739]
No “skip to main content” link is present on any of the application’s pages.  This means that keyboard-only users won’t be able to bypass any header content, a problem exacerbated by:

* different headers being used on different pages (for example between the Landing Page and main /app views); and
* the inconsistent (or missing) use of headings and ARIA landmark regions, as noted in component-specific tickets.

#### Suggested Solution 
 
A “Skip to main content” link should be present on every page, as the first control to receive keyboard focus and the first child element of the DOM.  When activated, it should move the user’s focus to the start of the main content, which should almost always be a level 1 heading.  In order to receive focus, a `tabindex` attribute with a value of `"-1"` should be added to that heading. 
 
Important note: the target of a skip link should never be set to an entire page section or container (such as a `<main>` or `<div>`), as a screen reader will attempt to voice all of the child content once it receives focus.  This will be annoying or confusing at best, but can also degrade performance as the software attempts to fill its speech queue and audio synthesis buffer with a large amount of data.

**Comment by alexrcabral** [comment]
Sina's suggestion is right on: "A “Skip to main content” link should be present on every page, as the first control to receive keyboard focus and the first child element of the DOM. When activated, it should move the user’s focus to the start of the main content, which should almost always be a level 1 heading. In order to receive focus, a tabindex attribute with a value of “-1” should be added to that heading."

Below are the defined headings by page 

Page: Landing Page
URL: https://mathshare.benetech.org/#/
Level 1 heading: Help students show and organize their math work

Page: Partnerships 
URL: https://mathshare.benetech.org/#/partners
Level 1 heading: "Partnerships" 

Page: Privacy
URL: https://mathshare.benetech.org/#/privacy
Level 1 heading: "Privacy Policy"

Page: Home Page (Viewing All Problem Sets) 
URL: https://mathshare.benetech.org/#/app/
Level 1 heading: "+New Problem Set" 
*This might change in the near future if we rearrange to show recent problem sets at the top of the page 

Page: Problem Set View 
URL (example): https://mathshare.benetech.org/#/app/problemSet/edit/FKIQS5XCUTC7Y
Level 1 heading: [Title of problem set] 

Page: Individual Problem View 
URL (example): https://mathshare.benetech.org/#/app/problem/edit/EVSOA5UGILPFU
Level 1 heading: [Title of problem] 
(as displayed in the orange ribbon)

**Comment by alexrcabral** [comment]
@johnhbenetech Is skip to main content meant to be a link or a button? 

Ours is currently coded as a button, and it sounds like Sina means for a link rather than a button (I know they're not necessarily interchangeable for screen readers). Both Google and Desmos use links for their Skip To Main functionality



Testing by tabbing through, it brings up skip to main content properly and navigates to the right spot on the Landing Page, the Home Page, and the Problem Set View. 

The skip to main content is not showing on the Partnerships page and the Privacy page, as well as the Individual Problem View. 

(Tested using VoiceOver on Chrome)

**Comment by johnhbenetech** [comment]
@rupeshparab can you check the above?

**Comment by rupeshparab** [comment]
I am adding the skip link to the Partnerships, Privacy and Individual page
The reason for keeping it a button is that the location to move the focus is dynamic, currently the focus is moved to the first element inside a container. The convention which I am following is I am using anchor where the location on click is static and I am using a button when the location on click is dynamic

**Comment by johnhbenetech** [comment]
@jscholes

**Comment by jscholes** [comment]
@rupeshparab - skip controls must be links in all instances, not buttons.

The problem with the current dynamic approach (via the focusOnMainContent function) is that it encourages pages not to have a level 1 heading.  That should never happen, so implementing a skip link can be as simple as:

* decide on an ID which will be shared by every level 1 heading across the site (e.g. "main-heading"); and then
* set that as the target of the skip link, everywhere.

    <a href="main-heading" ...>Skip to main content</a>
    ...
    <h1 id="main-heading" tabIndex="-1">...</h1>

This is a good pattern to follow, because it can highlight pages where maybe we still don't have a main level 1 heading even though we should.

**Comment by rupeshparab** [comment]
@jscholes 
One of the issues in using anchors to scroll to elements on page is that we are using a hash router in our app and the location in anchor will cause the react router to render NotFound page.
If we have to use anchors for skip link, we will need to change the router type which we are using currently (it will change the way the routes look like currently: https://mathshare-qa.diagramcenter.org/#/app/problemSet/solve/FQW3N7RPYNLI2 will change to https://mathshare-qa.diagramcenter.org/app/problemSet/solve/FQW3N7RPYNLI2).
It will make existing shared routes inaccessible. We can add redirection on basis of presence of hash when the page first loads.

Let me know if I should change the router

**Comment by jscholes** [comment]
@rupeshparab - could we not use an `<a>` with an `href` of `"javascript:void(0)"`, and then arrange scrolling and focus handling within the `onClick`?  I mean it's not ideal, but a link is 100 percent required here and it sounds like some work will be required to achieve it either way.

**Comment by rupeshparab** [comment]
@jscholes 
We can do this, but we will need to disable linter for this rule on the line: 'Script URL is a form of eval.eslint(no-script-url)'
In #747 we had tried to remove all of the linting rules disablers

**Comment by jscholes** [comment]
@rupeshparab - I'm okay with silencing this single linter comment just for one line of code.  #747 related to accessibility warnings in the linter which were actually helpful, because more often than not the behaviour on the described line was incorrect.

**Comment by rupeshparab** [comment]
@jscholes 
I have changed the skip link to an anchor

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/a47c35bdd4c23e73f327c6dc836d558baefc1950.
